Add proposal for low-level CSR access API

dongyongtao opened the pull request:
- the yaml file will be added when the parser supports macro-define-function type

Commit:
Add proposal for low-level CSR access API

- the yaml file will be added when the parser supports macro-define-function type

dongyongtao commented:
maybe there's a feature demand for csi_parser

https://github.com/riscv-non-isa/riscv-rvm-csi/issues/8#issuecomment-1724704143

fanghuaqi commented on api/C/include/csi_csr_access.h:
```diff
@@ -0,0 +1,151 @@
+/*
+ * macros
+ */
+#define __STR(s)                #s
```
This macro maybe need to provided as a generic one in rvm-csi, currently we just inline it in this header file

fanghuaqi commented on api/C/include/csi_csr_access.h:
```diff
@@ -0,0 +1,151 @@
+ */
+#define __STR(s)                #s
+#define STRINGIFY(s)            __STR(s)
+
```
Please define an extra macro `#define __ASM                        __asm`, which is used in below macro blocks

Commit:
Add yaml file of csi_csr_access.h

- update the csi_csr_access.h
- add __ASM define macro
- update this yaml filename into rvm-csi-spec.yaml

Signed-off-by: dongyongtao <dongyongtao@nucleisys.com>

fanghuaqi commented on api/C/include/csi_csr_access.h:
```diff
@@ -18,18 +18,13 @@
- * @param csr: CSR macro definition
- * @param val: value to store into the CSR register
- * @return : the CSR register value before written
+ * Read the content of csr register to __v, then write content of val into csr
```
Missing argument, it might not be able to generate doxygen doc like this,

![image](https://github.com/riscv-non-isa/riscv-rvm-csi/assets/1538922/a7a21164-bc07-46cb-8549-53e9442b75a6)

dongyongtao replied:
"@return : "  "@param csr:"   and" :  "is not allowed in this yaml,  so I have to omit these doxygen keywords.

WiseCrohn replied:
I can extend the schema to support a macro with documented arguments.  But let's treat that separately from this PR.

fanghuaqi commented:
Hi @WiseCrohn @VincentCui632 ， any feedback on this csr access api proposal

WiseCrohn commented:
I think most of this was taken from NMSIS https://github.com/Nuclei-Software/NMSIS/blob/master/NMSIS/Core/Include/core_feature_base.h.  In that file, the code is surrounded by `#ifndef __ASSEMBLY__` , in case it gets accidentally included from assembly code.  I think that's a good idea, but it might be better to use `#ifndef __ASSEMBLER__`.  I think `__ASSEMBLY__` is defined by linux builds, but `__ASSEMBLER__` is defined by gcc itself (https://gcc.gnu.org/onlinedocs/cpp/Standard-Predefined-Macros.html).

WiseCrohn commented:
We are proposing that RVM-CSI includes https://github.com/riscv/riscv-opcodes/blob/master/encoding.h, by making the riscv-opcodes repo a submodule and copying this file into our API.  This is useful to bring in all the definitions of CSR bitfields.  But this file includes versions of read_csr, write_csr, swap_csr, set_csr and clear_csr that make the equivalent versions in this file redundant?

fanghuaqi commented:
> We are proposing that RVM-CSI includes https://github.com/riscv/riscv-opcodes/blob/master/encoding.h, by making the riscv-opcodes repo a submodule and copying this file into our API. This is useful to bring in all the definitions of CSR bitfields. But this file includes versions of read_csr, write_csr, swap_csr, set_csr and clear_csr that make the equivalent versions in this file redundant?

I think we can define an API of csr operation, so it can cover different compiler difference. If this implementation are okay, the rvm-csi csr version can just define macro to reuse these API.

WiseCrohn commented:
@fanghuaqi @VincentCui632 I have submitted a PR #11 which allows to define a macro with parameters and return value. Also, an option "no-assembler" is introduced which causes C file contents to be surrounded by #ifndef __ASSEMBLER__.  If we merge this then it can help with the issues raised here above.

dongyongtao committed:
Merge branch 'riscv-non-isa:main' into main

Commit:
Update the yaml file supporting to define a macro with parameters and return value

- update the csi_ll_csr_access.h
- #ifndef __ASSEMBLER__ is introduced

Signed-off-by: dongyongtao <dongyongtao@nucleisys.com>

fanghuaqi reviewed (approved):
LGTM

WiseCrohn reviewed (approved):
Looks good, thanks
